src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs

```diff
@@ -41,45 +42,61 @@ public override void Initialize(AnalysisContext analysisContext)
             {
-                var formatInfo = new StringFormatInfo(compilationContext.Compilation);
+                var formatInfo = new StringFormatInfo(context.Compilation);
 
-                compilationContext.RegisterOperationAction(operationContext =>
+                context.RegisterOperationAction(context =>
                 {
-                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    var invocation = (IInvocationOperation)context.Operation;
 
-                    StringFormatInfo.Info? info = formatInfo.TryGet(invocation.TargetMethod, operationContext);
+                    StringFormatInfo.Info? info = formatInfo.TryGet(invocation.TargetMethod, context);
                     if (info == null || invocation.Arguments.Length <= info.FormatStringIndex)
                     {
                         // not a target method
                         return;
                     }
 
+                    if (info.ExpectedStringFormatArgumentCount >= 0 &&
+                        invocation.TargetMethod.IsVararg)
+                    {
+                        // __arglist is not supported here (see https://github.com/dotnet/roslyn/issues/7346)
+                        return;
+                    }
+
                     IArgumentOperation formatStringArgument = invocation.Arguments[info.FormatStringIndex];
-                    if (!object.Equals(formatStringArgument?.Value?.Type, formatInfo.String) ||
+                    if (!Equals(formatStringArgument?.Value?.Type, formatInfo.String) ||
                         !(formatStringArgument?.Value?.ConstantValue.Value is string))
                     {
                         // wrong argument
                         return;
                     }
 
                     var stringFormat = (string)formatStringArgument.Value.ConstantValue.Value;
-                    int expectedStringFormatArgumentCount = GetFormattingArguments(stringFormat);
+                    var stringFormatIndexes = GetStringFormatItemIndexes(stringFormat);
+                    int expectedStringFormatArgumentCount = stringFormatIndexes?.Count ?? -1;
 
-                    // explicit parameter case
-                    if (info.ExpectedStringFormatArgumentCount >= 0)
+                    // Validate string format item indexes
+                    if (expectedStringFormatArgumentCount > 0)
                     {
-                        // __arglist is not supported here
-                        if (invocation.TargetMethod.IsVararg)
+                        var missingIndexes = Enumerable.Range(0, stringFormatIndexes.Max())
+                            .Except(stringFormatIndexes);
+                        if (missingIndexes.Any())
                         {
-                            // can't deal with this for now.
+                            context.ReportDiagnostic(invocation.CreateDiagnostic(Rule));
```

Comment:
We may want to report with a different diagnostic (message + desc).

Reply:
I think, since we are saying we want to break with legacy FxCop behavior that introducing a new, more specific warning message would be nice.

---
In reply to: [495128627](https://github.com/dotnet/roslyn-analyzers/pull/4226#discussion_r495128627) [](ancestors = 495128627)

Reply:
I would still keep the same rule ID except if you think that's valuable to let users disable some part of the rule (I don't see the need as I don't expect much FPs).